code/game/machinery/computer/pod.dm

```diff
@@ -39,92 +56,80 @@
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "MassDriverControl", name)
+		ui.open()
+
+/obj/machinery/computer/pod/ui_data(mob/user)
+	var/list/data = list()
+	data["connected"] = connected ? TRUE : FALSE
+	data["seconds"] = round(time % 60)
+	data["minutes"] = round((time - data["seconds"]) / 60)
+	data["timing"] = timing
+	data["power"] = connected ? connected.power : 0.25
+	data["poddoor"] = FALSE
+	for(var/obj/machinery/door/poddoor/door in range(range, src))
+		if(door.id == id)
+			data["poddoor"] = TRUE
+			break
+	return data
+
+/obj/machinery/computer/pod/ui_act(action, list/params)
+	if(..())
 		return
-
-	var/dat = ""
-	if(connected)
-		var/d2
-		if(timing)	//door controls do not need timers.
-			d2 = "<A href='?src=[REF(src)];time=0'>Stop Time Launch</A>"
-		else
-			d2 = "<A href='?src=[REF(src)];time=1'>Initiate Time Launch</A>"
-		dat += "<HR>\nTimer System: [d2]\nTime Left: [DisplayTimeText((time SECONDS))] <A href='?src=[REF(src)];tp=-30'>-</A> <A href='?src=[REF(src)];tp=-1'>-</A> <A href='?src=[REF(src)];tp=1'>+</A> <A href='?src=[REF(src)];tp=30'>+</A>"
-		var/temp = ""
-		var/list/L = list( 0.25, 0.5, 1, 2, 4, 8, 16 )
-		for(var/t in L)
-			if(t == connected.power)
-				temp += "[t] "
-			else
-				temp += "<A href = '?src=[REF(src)];power=[t]'>[t]</A> "
-		dat += "<HR>\nPower Level: [temp]<BR>\n<A href = '?src=[REF(src)];alarm=1'>Firing Sequence</A><BR>\n<A href = '?src=[REF(src)];drive=1'>Test Fire Driver</A><BR>\n<A href = '?src=[REF(src)];door=1'>Toggle Outer Door</A><BR>"
-	else
-		dat += "<BR>\n<A href = '?src=[REF(src)];door=1'>Toggle Outer Door</A><BR>"
-	dat += "<BR><BR><A href='?src=[REF(user)];mach_close=computer'>Close</A>"
-	add_fingerprint(usr)
-	var/datum/browser/popup = new(user, "computer", title, 400, 500)
-	popup.set_content(dat)
-	popup.open()
-
-/obj/machinery/computer/pod/process()
-	if(!..())
+	if(!allowed(usr))
+		to_chat(usr, "<span class='warning'>Access denied.</span>")
 		return
-	if(timing)
-		if(time > 0)
-			time = round(time) - 1
-		else
-			alarm()
-			time = 0
-			timing = 0
-		updateDialog()
 
-
-/obj/machinery/computer/pod/Topic(href, href_list)
-	if(..())
-		return
-	if(usr.contents.Find(src) || (in_range(src, usr) && isturf(loc)) || issilicon(usr))
-		usr.set_machine(src)
-		if(href_list["power"])
-			var/t = text2num(href_list["power"])
-			t = min(max(0.25, t), 16)
-			if(connected)
-				connected.power = t
-		if(href_list["alarm"])
+	switch(action)
+		if("set_power")
+			if(!connected)
+				return
+			var/value = text2num(params["power"])
+			if(!value)
+				return
+			value = clamp(value, 0.25, 16)
+			connected.power = value
+			return TRUE
+		if("launch")
 			alarm()
-		if(href_list["time"])
-			timing = text2num(href_list["time"])
-		if(href_list["tp"])
-			var/tp = text2num(href_list["tp"])
-			time += tp
-			time = min(max(round(time), 0), 120)
-		if(href_list["door"])
+			return TRUE
+		if("time")
+			timing = !timing
+			return TRUE
```

Comment:
This is a very good place to put a bit of logic to start or stop processing.

Reply:
Resolved.